### PR TITLE
Implement new property #updateLocationInBackground

### PR DESCRIPTION
**Pull request by christocracy**
This property defaults to `true` (the current behaviour).  When set to `false` (with `showsUserLocation` set to `true`), the component will not update the location while app is in background.

This allows 3rd-party location-manager modules [such as mine](http://transistorsoft.github.io/react-native-background-geolocation/) to better handle background geolocation, since a View component should not be responsible for updating location while app is running in background.

### CODE REVIEW

File: RCTMapboxGL.ios.js [hunk 1]
```diff
@@ -4,6 +4,9 @@ var React = require('react-native');
 var { NativeModules, requireNativeComponent } = React;
 
 var MapMixins = {
+  _findHandle(mapRef) {
+    return React.findNodeHandle(this.refs[mapRef]);
+  },
   setDirectionAnimated(mapRef, heading) {
```
Comment (bsudekum) [inline RCTMapboxGL.ios.js:10]: @christocracy why was this helper function added? Can't this be accomplished in the same manner as the rest as mixin functions? IE 

```
NativeModules.MapboxGLManager.addAnnotations(React.findNodeHandle(this.refs[mapRef]), annotations);
```

Comment (christocracy) [inline RCTMapboxGL.ios.js:10]: I see a lot of duplicated code in the API methods.  I think each method should use the helper.

**Comment by bsudekum** [comment]
While testing I set `setUpdateLocationInBackground={true}`, then set simulator location to `City bike ride`. I then closed the app and saw no location updates when logging out from the `onUpdateUserLocation` event handler. @christocracy Is this expected?

**Comment by christocracy** [comment]
I'll have a look tonight

On Wednesday, August 12, 2015, Bobby Sudekum notifications@github.com
wrote:

> While testing I set setUpdateLocationInBackground={true}, then set
> simulator location to City bike ride. I then closed the app and saw no
> location updates when logging out from the onUpdateUserLocation event
> handler. @christocracy https://github.com/christocracy Is this
> expected?
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/mapbox/react-native-mapbox-gl/pull/78#issuecomment-130474713
> .

## 

Snet form Gmail Mobile

**Comment by christocracy** [comment]
You've used the _method-name_ `setUpdateLocationInBackground` -- the property's name is `updateLocationInBackground`.

```
      <MapboxGLMap
          showsUserLocation={true}
          updateLocationInBackground={true} />
```

Confirmed working.

**Comment by bsudekum** [comment]
@christocracy ah good catch. I'm not getting the following error when resuming the app intermittently:

```
-[__NSCFType onSuspend:]: unrecognized selector sent to instance 0x7fa70c3efea0
2015-08-13 09:06:59.864 test2[75245:6368478] *** Terminating app due to uncaught exception 'NSInvalidArgumentException', reason: '-[__NSCFType onSuspend:]: unrecognized selector sent to instance 0x7fa70c3efea0'
*** First throw call stack:
(
    0   CoreFoundation                      0x000000010ef0aa75 __exceptionPreprocess + 165
    1   libobjc.A.dylib                     0x000000010e38ebb7 objc_exception_throw + 45
    2   CoreFoundation                      0x000000010ef11d1d -[NSObject(NSObject) doesNotRecognizeSelector:] + 205
    3   CoreFoundation                      0x000000010ee699dc ___forwarding___ + 988
    4   CoreFoundation                      0x000000010ee69578 _CF_forwarding_prep_0 + 120
    5   CoreFoundation                      0x000000010eeda7fc __CFNOTIFICATIONCENTER_IS_CALLING_OUT_TO_AN_OBSERVER__ + 12
    6   CoreFoundation                      0x000000010edda204 _CFXNotificationPost + 2484
    7   Foundation                          0x000000010dee9df8 -[NSNotificationCenter postNotificationName:object:userInfo:] + 66
    8   UIKit                               0x0000000110161aa7 -[UIApplication _handleApplicationDectivationWithScene:shouldForceExit:transitionContext:completion:] + 1054
    9   UIKit                               0x000000011016e833 __88-[UIApplication _handleApplicationLifecycleEventWithScene:transitionContext:completion:]_block_invoke + 87
    10  UIKit                               0x000000011016e7cd -[UIApplication _handleApplicationLifecycleEventWithScene:transitionContext:completion:] + 349
    11  UIKit                               0x000000011015945a -[UIApplication scene:didUpdateWithDiff:transitionContext:completion:] + 486
    12  FrontBoardServices                  0x000000011431a243 __31-[FBSSerialQueue performAsync:]_block_invoke + 16
    13  CoreFoundation                      0x000000010ee3fc7c __CFRUNLOOP_IS_CALLING_OUT_TO_A_BLOCK__ + 12
    14  CoreFoundation                      0x000000010ee359c5 __CFRunLoopDoBlocks + 341
    15  CoreFoundation                      0x000000010ee35183 __CFRunLoopRun + 851
    16  CoreFoundation                      0x000000010ee34bc6 CFRunLoopRunSpecific + 470
    17  GraphicsServices                    0x0000000112157a58 GSEventRunModal + 161
    18  UIKit                               0x000000011015b580 UIApplicationMain + 1282
    19  test2                               0x000000010ce81fff main + 111
    20  libdyld.dylib                       0x000000011156c145 start + 1
    21  ???                                 0x0000000000000001 0x0 + 1
)
libc++abi.dylib: terminating with uncaught exception of type NSException
(lldb) 
```

Nor am I seeing the location logged to the console while the app is backgrounded:



Also, want to update the mixin code so all methods use `_findHandle`?

**Comment by christocracy** [comment]
ahh, right.  I forgot to detach `NSNotification` listeners in `dealloc`.  Simply need to add following.

```
- (void)dealloc
{
    [[NSNotificationCenter defaultCenter] removeObserver:self];
}
```

I'll update all methods to use `#_findHandle`.

**Comment by bsudekum** [comment]
:+1:

**Comment by bsudekum** [comment]
@christocracy I pulled down your recent work and I'm still seeing both the bug above and no locations sent to console while the app is backgrounded.

**Comment by christocracy** [comment]
You won't see background location updates until you enable `location` background-mode.  Enable that.



As for the bug,  I cannot duplicate that.  I _have_ experienced this issue _in the past_ and clearly the [solution is](http://stackoverflow.com/questions/6498792/nsnotification-leads-to-segmentation-fault):

```
- (void)dealloc
{
    [[NSNotificationCenter defaultCenter] removeObserver:self];
}
```

I booted your Example app from scratch and everything is fine here.

**Pull request closed without merging**